### DISCUSSION TITLE
Using the source file as testsfile (Byte Tester)

### DISCUSSION THREAD

**Discussion by lhsazevedo** [vhelin/wla-dx#405] [Ideas]
I was thinking that it would be cool if the Byte Tester could read the test data from comments an source file, instead of an separate testsfile.

## How we do now

Right now we have to write the assembly file, and a separate testsfile with the test instructions. It's great, but it's also a bit hard to maintain and to create new ones, because we have to context-switch between the two files. Also, there's no visible info for developers of which bytes each instruction should generate.
  
For example, look at the z80/mnemonics_test assembly file:
```assembly
.db "01>"
res 0, (ix+$12), d // $DD $CB $12 $82
res 0,(iy+$69),a   // $FD $CB $69 $87
.db "<01"

.db "02>"
res 3, (ix+$34), h // $DD $CB $34 $9C
res 3, (iy+$56)    // $FD $CB $56 $9E
.db "<02"
```

And its testsfile:
```
result.rom
TEST-01 01 START DD CB 12 82 FD CB 69 87 END
TEST-02 02 START DD CB 34 9C FD CB 56 9E END
```

Note that the the resulting bytes are already commented out in the assembly file, right next to the opcodes that should generate them.

## Proposal

Add a flag to the Byte Tester indicating that it should read the test instructions from comments of an assembly file.
It would look for comments starting with an maker (I used `@BT` here as suggestion), and read instructions from it.

That way we could write the z80 test mentioned earlier the following way:
```assembly
; @BT result.rom
.db "01>"           ; @BT TEST-01 01 START
res 0, (ix+$12), d  ; @BT DD CB 12 82
res 0,(iy+$69),a    ; @BT FD CB 69 87
.db "<01"           ; @BT END

; Byte Tester will ignore this comment

.db "02>"           ; @BT TEST-02 02 START
res 3, (ix+$34), h  ; @BT DD CB 34 9C
res 3, (iy+$56)     ; @BT FD CB 56 9E
.db "<02"           ; @BT END
```

Note that the placement of the comments is completely irrelevant. We could write the same test as: 
```assembly
; @BT result.rom
; @BT TEST-01 01 START
; @BT DD CB 12 82
; @BT FD CB 69 87
; @BT END

.db "01>"
res 0, (ix+$12), d
res 0, (iy+$69),a
.db "<01"

; @BT TEST-02 02 START
; @BT DD CB 34 9C
; @BT FD CB 56 9E
; @BT END

.db "02>"
res 3, (ix+$34), h
res 3, (iy+$56)
.db "<02"
```

or

```assembly
; @BT result.rom
; @BT TEST-01 01 START
; @BT DD CB 12 82
; @BT FD CB 69 87
; @BT END
; @BT TEST-02 02 START
; @BT DD CB 34 9C
; @BT FD CB 56 9E
; @BT END

.db "01>"
res 0, (ix+$12), d
res 0,(iy+$69),a
.db "<01"

.db "02>"
res 3, (ix+$34), h
res 3, (iy+$56)
.db "<02"
```

**Comment by vhelin** [comment]
Good idea and proposal! :)

**Reply by cr1901** [reply]
This is analogous to what LLVM does for it's own test suite. Can't imagine<sup>1</sup> it would be difficult to add this functionality either...

1. Famous last words.

**Comment by lhsazevedo** [comment]
Started working on it [here](https://github.com/vhelin/wla-dx/compare/master...lhsazevedo:byte-tester-comments) but is very, very WIP. Still need to replace the `fscanf` with something that reads from the input string, and test the flag parsing loop.

**Comment by vhelin** [comment]
A question: Currently it seems to support comments that begin with a ';' ("; @BT") - could it also support comments beginning with "//" and "/*"?

**Reply by lhsazevedo** [reply]
Yes, it shouldn't be too difficult. For the block comment `/*`, I guess we can seek for an `*/` as terminator instead of a line feed `\n', as used in the other comment types.

I'm finally understanding pointers \o/. This [C89 draft](http://port70.net/~nsz/c/c89/c89-draft.html) is helping a lot.

**Reply by lhsazevedo** [reply]
Added it, branch updated. Now we only need to parse from the input string instead of the file stream.

**Reply by vhelin** [reply]
If string parser is too much work you could always save the string to a file... :)

**Reply by lhsazevedo** [reply]
Yeah, this is my plan B.

I read that you can use "%n" on `sscanf` to get the number of characters. I guess this can replace `fscanf` pretty easily.

**Comment by lhsazevedo** [comment]
Done, [branch](https://github.com/vhelin/wla-dx/compare/master...lhsazevedo:byte-tester-comments) updated.

Next steps:
- Bike-shedding time! :bike: We need to choose the comment token. I started with `@BT` without thinking too much.
As pointed out by @cr1901, LLVM uses¹ `CHECK:`. Wiz uses² `BLOCK`.
- :bike: Choose the flag to pass to Byte Tester, I started with `-s`.
-  Figure out how the test runner can be updated to run tests using the new format, as there's now testsfile and Byte Tester needs an extra flag.
https://github.com/vhelin/wla-dx/blob/c71c069a1dea536aed43d7875027b407ddb00b1b/run_tests.sh#L16-L20

---

1. https://github.com/llvm/llvm-project/blob/e8bce83996313ed3f4f5fce43107530d49fc3b64/llvm/test/Analysis/AliasSet/argmemonly.ll
2. https://github.com/wiz-lang/wiz/blob/972cd9dfcd52702ade90d90d9ec5eaa049753627/tests/block/6502_alu.wiz

**Reply by lhsazevedo** [reply]
Then I'll open an PR for code review. I can do this now if you prefer.

**Reply by vhelin** [reply]
I'm ok with @BT and -s

**Comment by lhsazevedo** [comment]
PR created #406
